### PR TITLE
Update int8_simd.h

### PR DESCRIPTION
**Pull request by JerryShih**
If we use clang with sse4 support, we will have the function redefinition
error between [1] and [2]. This patch try to add some checkings to fix this
problem.

I just turn on USE_NATIVE_ARCH with clang, then I hit the redefinition error.

[1]
caffe2/operators/quantized/int8_simd.h
[2]
third_party/gemmlowp/gemmlowp/fixedpoint/fixedpoint_sse.h

### CODE REVIEW

**Comment by JerryShih** [comment]
@Maratyszcza 
Could you please check this patch?

File: caffe2/operators/quantized/int8_simd.h [hunk 1]
```diff
@@ -19,10 +19,11 @@
 
 #if defined(__SSE4_2__) && defined(__clang__)
 #define INT8_NEON_SIMD
+#include "NEON_2_SSE.h"
```
Comment (JerryShih) [inline caffe2/operators/quantized/int8_simd.h]: I don't see the neon2sse.h file in pytorch repo. Instead, we have ./third_party/neon2sse/NEON_2_SSE.h in repo.

File: caffe2/operators/quantized/int8_simd.h [hunk 2]
```diff
@@ -19,10 +19,11 @@
...
+#endif
 
-#include "neon2sse.h"
+#if !defined(GEMMLOWP_SSE4) && !defined(GEMMLOWP_NEON)
```
Comment (JerryShih) [inline caffe2/operators/quantized/int8_simd.h]: The following functions(e.g. Dup() and BitAnd()) are already in GEMMLOWP_NEON and GEMMLOWP_SSE4 header.

File: caffe2/operators/quantized/int8_simd.h (hunk 1 above)
Comment (Maratyszcza) [reply caffe2/operators/quantized/int8_simd.h]: I am puzzled as to how it worked before

**Comment by Maratyszcza** [comment]
@JerryShih I changed `neon2sse.h` include to `NEON_2_SSE.h` in 6d9a7d0e60affa3ed32e19490a57f5d829714f41, please rebase on top of ti

**Comment by JerryShih** [comment]
@Maratyszcza 
Here is the rebased commit.

**Comment by ezyang** [comment]
Unfortunately, this seems to break our internal build for some reason. The error looks like this:

```
[redacted]/build/gemmlowp/include/gemmlowp/fixedpoint/fixedpoint.h:434:35: error: no type named 'ScalarRawType' in 'gemmlowp::FixedPointRawTypeTraits<__attribute__((__vector_size__(2 * sizeof(long long)))) long long>'
  typedef typename RawTypeTraits::ScalarRawType ScalarRawType;
          ~~~~~~~~~~~~~~~~~~~~~~~~^~~~~~~~~~~~~
[redacted]/build/gemmlowp/include/gemmlowp/fixedpoint/fixedpoint.h:434:35: error: no type named 'ScalarRawType' in 'gemmlowp::FixedPointRawTypeTraits<__attribute__((__vector_size__(2 * sizeof(long long)))) long long>'
  typedef typename RawTypeTraits::ScalarRawType ScalarRawType;
          ~~~~~~~~~~~~~~~~~~~~~~~~^~~~~~~~~~~~~
[redacted]/build/gemmlowp/include/gemmlowp/fixedpoint/fixedpoint.h:468:20: error: no matching function for call to 'Dup'
    retval.raw() = Dup<RawType>(x);
                   ^~~~~~~~~~~~
[redacted]/build/gemmlowp/include/gemmlowp/fixedpoint/fixedpoint.h:141:33: error: cannot initialize a variable of type '__attribute__((__vector_size__(2 * sizeof(long long)))) long long const' (vector of 2 'long long' values) with an rvalue of type 'int'
  static constexpr tIntegerType zero = 0;
                                ^      ~
[redacted]/build/gemmlowp/include/gemmlowp/fixedpoint/fixedpoint.h:149:38: error: invalid argument type '__attribute__((__vector_size__(2 * sizeof(long long)))) long long' (vector of 2 'long long' values) to unary expression
  return MaskIfNonZero<tIntegerType>(!a);
                                     ^~
[redacted]/build/gemmlowp/include/gemmlowp/fixedpoint/fixedpoint.h:434:35: error: no type named 'ScalarRawType' in 'gemmlowp::FixedPointRawTypeTraits<__attribute__((__vector_size__(2 * sizeof(long long)))) long long>'
  typedef typename RawTypeTraits::ScalarRawType ScalarRawType;
          ~~~~~~~~~~~~~~~~~~~~~~~~^~~~~~~~~~~~~
[redacted]/build/gemmlowp/include/gemmlowp/fixedpoint/fixedpoint.h:214:3: error: static_assert failed due to requirement 'std::is_same<__attribute__((__vector_size__(2 * sizeof(long long)))) long long, void>::value' "unimplemented"
  static_assert(std::is_same<IntegerType, void>::value, "unimplemented");
  ^             ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
[redacted]/build/gemmlowp/include/gemmlowp/fixedpoint/fixedpoint.h:468:20: error: no matching function for call to 'Dup'
    retval.raw() = Dup<RawType>(x);
                   ^~~~~~~~~~~~
[redacted]/build/gemmlowp/include/gemmlowp/fixedpoint/fixedpoint.h:468:20: error: no matching function for call to 'Dup'
    retval.raw() = Dup<RawType>(x);
                   ^~~~~~~~~~~~
[redacted]/build/gemmlowp/include/gemmlowp/fixedpoint/fixedpoint.h:339:28: error: no matching function for call to 'Dup'
  const IntegerType mask = Dup<IntegerType>((1ll << exponent) - 1);
                           ^~~~~~~~~~~~~~~~
[redacted]/build/gemmlowp/include/gemmlowp/fixedpoint/fixedpoint.h:340:28: error: no matching function for call to 'Dup'
  const IntegerType zero = Dup<IntegerType>(0);
                           ^~~~~~~~~~~~~~~~
[redacted]/build/gemmlowp/include/gemmlowp/fixedpoint/fixedpoint.h:341:27: error: no matching function for call to 'Dup'
  const IntegerType one = Dup<IntegerType>(1);
                          ^~~~~~~~~~~~~~~~
[redacted]/build/gemmlowp/include/gemmlowp/fixedpoint/fixedpoint.h:339:28: error: no matching function for call to 'Dup'
  const IntegerType mask = Dup<IntegerType>((1ll << exponent) - 1);
                           ^~~~~~~~~~~~~~~~
[redacted]/build/gemmlowp/include/gemmlowp/fixedpoint/fixedpoint.h:340:28: error: no matching function for call to 'Dup'
  const IntegerType zero = Dup<IntegerType>(0);
                           ^~~~~~~~~~~~~~~~
[redacted]/build/gemmlowp/include/gemmlowp/fixedpoint/fixedpoint.h:341:27: error: no matching function for call to 'Dup'
  const IntegerType one = Dup<IntegerType>(1);
                          ^~~~~~~~~~~~~~~~
[redacted]/build/gemmlowp/include/gemmlowp/fixedpoint/fixedpoint.h:339:28: error: no matching function for call to 'Dup'
  const IntegerType mask = Dup<IntegerType>((1ll << exponent) - 1);
                           ^~~~~~~~~~~~~~~~
[redacted]/build/gemmlowp/include/gemmlowp/fixedpoint/fixedpoint.h:340:28: error: no matching function for call to 'Dup'
  const IntegerType zero = Dup<IntegerType>(0);
                           ^~~~~~~~~~~~~~~~
[redacted]/build/gemmlowp/include/gemmlowp/fixedpoint/fixedpoint.h:341:27: error: no matching function for call to 'Dup'
  const IntegerType one = Dup<IntegerType>(1);
                          ^~~~~~~~~~~~~~~~
```

**Comment by JerryShih** [comment]
What's the configuration for this build? Could I get the docker image to reproduce the error?
@ezyang

**Comment by JerryShih** [comment]
And the error line messages doesn't match the recent pytorch master branch code base. :(

**Comment by ezyang** [comment]
@JerryShih Unfortunately, there isn't a Docker image for this build. I was hoping that the error would be obvious from inspection. Let me rerun the build and get you a newer error trace.

**Comment by ezyang** [comment]
Oh, I bet the problem is that the version of gemmlowp internally is too old.

**Comment by ezyang** [comment]
paging @harouwu

**Pull request closed without merging**